### PR TITLE
Add Estimated Delivery Date to Checkout Attributes

### PR DESCRIPTION
**Pull request by danielcliu-ifixit**
## Overview

Basically just replicate the functionality from `shopify-v2's`[ checkout-attributes.ts](https://github.com/iFixit/shopify-v2/blob/80abea2a1865cbab9ae79d4929e26d7d7bd4a9e6/src/scripts/checkout/checkout-attributes.ts#L120)

Im not 100% sure this code is fullproof but we seem to have been using for a long time so its probably fine
Should set the estimate to 1 week if there is no estimated data.

## QA

Make sure that the estimated date shows up in the order's attributes, (similar to #54) and its accurate. If the shipping doesn't have an estimate attached to it, the estimate should be 1 week. To check this, you will probably need to set your delivery address to somewhere in europe, ive been using germany.

Closes: #40

### CODE REVIEW

**Comment by danielcliu-ifixit** [comment]
dev_block :v:  on #54 going out first

File: extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx [hunk 1]
```diff
@@ -1,14 +1,112 @@
...
       <BlockStack>
          <CheckoutHint />
          <VatSection />
       </BlockStack>
-   )
-);
+   );
+}
+
+function getDefaultEstimatedDeliveryDate() {
```
Comment (danielcliu-ifixit) [inline extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx:60]: based on https://github.com/iFixit/shopify-v2/blob/4e1c1341fb89d66c697f934b6b9b127e0be1ae47/src/scripts/checkout/checkout-attributes.ts#L103

File: extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx [hunk 2]
```diff
@@ -1,14 +1,112 @@
...
+      getDefaultEstimatedDeliveryDate()
+   );
+
+   useEffect(() => {
+      const newDeliveryGroup =
+         deliveryGroupTarget?.deliveryOptions?.find(
+            (option) =>
+               option.handle ===
+               deliveryGroupTarget?.selectedDeliveryOption?.handle
+         ) ?? null;
```
Comment (masonmcelvain) [inline extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx:30]: It seems like [`useShippingOptionTarget`](https://shopify.dev/docs/api/checkout-ui-extensions/2024-01/apis/delivery#useShippingOptionTarget) would meet our needs here and be simpler.

File: extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx [hunk 3]
```diff
@@ -1,14 +1,112 @@
...
+function getDefaultEstimatedDeliveryDate() {
+   // One week from now
+   const estimatedDeliveryDate = new Date(Date.now() + 7 * 24 * 60 * 60 * 1000);
+   // Format: YYYY-MM-DD
+   return estimatedDeliveryDate.toISOString().split('T')[0];
+}
+
+function getEstimatedDeliveryDate(secondsToDelivery: number): string {
+   const maxDays = Math.ceil(secondsToDelivery / 60 / 60 / 24);
+   const daysUntilDelivery = workDaysToDaysFromNow(maxDays);
```
Comment (masonmcelvain) [inline extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx:69]: I don't think we can re-use `workDaysToDaysFromNow` from shopify-v2 here. In shopify-v2, we parse the number of business days from the DOM. Here, we are dealing with an absolute number of seconds.

Instead we need to convert the number of seconds into a corresponding date in the future. We don't need to worry about "business days" as far as I can tell.

File: extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx [hunk 4]
```diff
@@ -1,14 +1,112 @@
...
+         daysToAdd += 2 * Math.floor(remainingWorkDays / 5);
+         10;
+         // Exclude final weekend if remainingWorkDays resolves to an exact number of weeks
+         if (remainingWorkDays % 5 === 0) {
+            daysToAdd -= 2;
+         }
+      }
+   }
+   return daysToAdd;
+}
```
Comment (masonmcelvain) [inline extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx:106]: I would probably do something like:

```suggestion
function oneWeekFromNow() {
   return formatDate(new Date(Date.now() + 7 * 24 * 60 * 60 * 1000));
}

function estimatedDeliveryDate(deliveryEstimate: DeliveryEstimate) {
   return formatDate(new Date(
      Date.now() + deliveryEstimate.timeInTransit.upper * 1000
   ));
}

/** Format: YYYY-MM-DD */
function formatDate(date: Date) {
   return date.toISOString().split('T')[0];
}
```

**Review by masonmcelvain — APPROVED** [review]
CR 👏🏻 Nice work thanks @danielcliu-ifixit!

dev_block 👍🏻 on some potential simplifications

File: extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx (hunk 2 above)
Comment (danielcliu-ifixit) [reply extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx:30]: Im not sure i agree, that seems to be a way to target each individual shipping option target, but not all of them at once. Im not sure it makes sense to add, for example, to add a `purchase.checkout.shipping-option-item.render-after` just for this functionality, when we have the functionality here

File: extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx (hunk 3 above)
Comment (danielcliu-ifixit) [reply extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx:69]: for whatever reason, the shipping transit times appear to still be measure in business days, see https://admin.shopify.com/store/ifixit-checkout-dev/settings/shipping/profiles/115323928861

Comment (masonmcelvain) [reply extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx:69]: I see, it's confusing that the docs don't mention it. I think we should test it out on one of those carriers. i.e. log the transit time in seconds (ideally for an option that would ship over the weekend) and determine whether it's "business seconds" or absolute seconds.

File: extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx (hunk 2 above)
Comment (masonmcelvain) [reply extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx:30]: You're right, it's for a different target.

File: extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx [hunk 5]
```diff
@@ -1,14 +1,112 @@
...
+
+function isShippingOption(
+   option: DeliveryOption | null
+): option is ShippingOption {
+   return option?.type === 'shipping' || option?.type === 'pickup';
+}
```
Comment (masonmcelvain) [inline extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx:112]: Another thing I noticed just now is that the docs say the type for a `ShippingOption` is `shipping` or `local`. `pickup` looks like it's for `PickupLocationOption`.

image

Comment (danielcliu-ifixit) [reply extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx:112]: oh shoot youre right

File: extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx (hunk 3 above)
Comment (danielcliu-ifixit) [reply extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx:69]: 2-4 business days converted today to
```
{
    "lower": 172800,
    "upper": 345600
}
```

so i think its just 2-4 days in seconds and then we have to account for any non-business days in our estimate

Comment (masonmcelvain) [reply extensions/ui-extension/src/targets/purchase/checkout/shipping-option-list/render-after.tsx:69]: Thank you for doing that experiment so fast! I'm convinced 👍🏻

**Review by masonmcelvain — DISMISSED** [review]
CR 👏🏻

**Comment by masonmcelvain** [comment]
un_dev_block 👍🏻

**Comment by danielcliu-ifixit** [comment]
@kthaler im not sure exactly why youre gettting that descrepency, but ive changed the code so that it should be easier to replicate bugs if they continue. 
I cant recreate your issue, so un_deploy_block :v: 

Also if your order is over $50+ in germany you get two delivery options, that should be tested.

**Comment by masonmcelvain** [comment]
> the approx delivery date is 1 to 2 days, but the actual calculated shipping date says 5 days

Good observation @kthaler! I am not super concerned about it because this is a development store that may not have been configured exactly right. I think as long as the order attribute's estimated delivery date matches what is shown during checkout, then this code is doing what we expect it to. It sounds like that is the case based on your QA, so I'll un_deploy_block 👍🏻

**Review by masonmcelvain — DISMISSED** [review]
CR 👍🏻 though I don't understand the intent of the change in ebe3ce92ca4eec8103f23e201639dbebf57abe86

**Comment by danielcliu-ifixit** [comment]
@masonmcelvain im not sure why @kthaler got his issue, but its hard to test the `Date.now()` functionality because its based on data. this at least its easier to manipulate `currentDate` to replicate the error. At least thats how i was thinking about it

**Review by masonmcelvain — APPROVED** [review]
CR 👍🏻

**Comment by kthaler** [comment]
QA 💎 - The expected dates look good! Tested the case where the order value is over $50, then you get 2 shipping options. This is shipping to an address in Germany, but I also tested shipping to an address in Belgium and California. In all cases, the expected date matches the range given in the shipping step of checkout. 

| Standard    | Express |
| ----------- | ----------- |
| image     | image     |
| image   | image        |

**Pull request merged**